### PR TITLE
Add an option to pin to gpu for all estimators

### PR DESCRIPTION
**Pull request by Tixxx**
Signed-off-by: TJ <tix@uber.com>

## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description
Sometimes the estimator can be created on a GPU host which initializes model weights on one device. Pinning to GPU again in remote trainer can have strange behaviors such as "Variable resource not found" error for Tensorflow since tf thinks the model weights were initialized on a different device.

Fixes # (issue).
https://github.com/horovod/horovod/issues/3524
## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Review by chongxiaoc — CHANGES_REQUESTED** [review]
Generic param of Spark Estimator can be put in `EstimatorParams `, and also add Getter and Setter there.
https://github.com/horovod/horovod/blob/master/horovod/spark/common/params.py#L56

**Review by chongxiaoc — APPROVED** [review]
LGTM

**Comment by chongxiaoc** [comment]
@EnricoMi Can you help do a sanity check as well?

**Comment by chongxiaoc** [comment]
@EnricoMi Hold this PR unmerge for now.
I just synced with @Tixxx and we need to wait for Ray Tune to test this PR for integration test.

**Comment by EnricoMi** [comment]
> Hold this PR unmerge for now.

I have set this to draft, press "Ready for review" when you are happy to see this merged.

File: horovod/spark/common/params.py
Comment (EnricoMi) [inline horovod/spark/common/params.py:104]: indentation is a bit off here

**Comment by Tixxx** [comment]
The [fix](https://github.com/ray-project/ray/pull/22708) for the horovod incompatibility on ray tune's side was checked in just now. I'm running the integration test now and will merge it once that's done.

File: horovod/spark/keras/remote.py
Comment (EnricoMi) [inline horovod/spark/keras/remote.py:120]: no `f` needed

File: horovod/spark/keras/remote.py
Comment (EnricoMi) [inline horovod/spark/keras/remote.py:124]: same here

File: horovod/spark/lightning/remote.py
Comment (EnricoMi) [inline horovod/spark/lightning/remote.py:201]: It should be mentioned in the API that `pin_gpu==False` disables CUDA.

Here should be a **warning** that though CUDA is available is is not used as pin_gpu is false. Would otherwise be hard to find out why CUDA is not used.

File: horovod/spark/lightning/remote.py
Comment (EnricoMi) [inline horovod/spark/lightning/remote.py:199]: Why not using the logger? Why is there `verbose` when there is a logger?

**Review by EnricoMi — COMMENTED** [review]
Some minor improvements possible.

File: horovod/spark/lightning/remote.py
Comment (Tixxx) [reply horovod/spark/lightning/remote.py:201]: makes sense, added a check to emit message when cuda is available but pin_gpu=false

File: horovod/spark/keras/remote.py
Comment (Tixxx) [reply horovod/spark/keras/remote.py:120]: good catch

File: horovod/spark/common/params.py
Comment (Tixxx) [reply horovod/spark/common/params.py:104]: updated.

File: horovod/spark/lightning/remote.py
Comment (Tixxx) [reply horovod/spark/lightning/remote.py:199]: The logger doesn't write to stdout and stderr properly in this function since it's run in ray executor. The train_logger is for passing some specialized loggers(not writing to stdout and stderr directly) to pytorch lightning. I have tried using some generic logger modules here, they either failed to serialize or no output.

File: horovod/spark/torch/remote.py
Comment (EnricoMi) [inline horovod/spark/torch/remote.py:141]: The last bit of the condition has to go:
```suggestion
        cuda_available = torch.cuda.is_available()
```

File: horovod/spark/torch/remote.py
Comment (EnricoMi) [inline horovod/spark/torch/remote.py:145]: identation

File: horovod/spark/common/params.py
Comment (EnricoMi) [inline horovod/spark/common/params.py:106]: If not pinning to GPU means not using a GPU at all, then this should be called `use_gpu`. I thought "pin" has a subtle different meaning than "use".

Then, `use_gpu=False` becomes clear what it means even if GPU is available.

I would still keep the warning prints below.

Comment (Tixxx) [reply horovod/spark/common/params.py:106]: i see, then i think we can rename it to use_gpu

Comment (Tixxx) [reply horovod/spark/common/params.py]: renamed

**Comment by Tixxx** [comment]
some builds are failing with 
```
#6 4.879 W: GPG error: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4B469963BF863CC
  #6 4.879 E: The repository 'https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease' is not signed.
```

@EnricoMi  Do you know if this is because nv's key expired or we need to refresh on our side?

**Comment by Tixxx** [comment]
Integration test passed on my side. Changing this back to ready state.

**Comment by EnricoMi** [comment]
> @EnricoMi Do you know if this is because nv's key expired or we need to refresh on our side?

@Tixxx This is something NVidia has to fix on their side.

@romerojosh @nvcastet NVidia's CUDA docker image cannot `apt update` because one of NVidia's repos' key is invalid:

```
docker run --rm -it nvidia/cuda:10.0-devel-ubuntu18.04 apt-get update
...
Err:9 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease                        
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4B469963BF863CC
```

Something you can escalate?

**Review by EnricoMi — APPROVED** [review]
LGTM! Please wait with merge for fixed CI.

**Comment by Tixxx** [comment]
Looks like nvidia rolled out a[ rotating key mechanism](https://forums.developer.nvidia.com/t/notice-cuda-linux-repository-key-rotation/212771). I will try to fix it by downloading the corresponding keys.

**Comment by EnricoMi** [comment]
@Tixxx this is how @tgaddair fixed it for Ludwig: https://github.com/ludwig-ai/ludwig/commit/1a4f67914bc3894c7ada67a5b985c9c1ca2a3cc0

**Comment by EnricoMi** [comment]
@Tixxx looks like your fix works!

**Comment by Tixxx** [comment]
> 

yea, I tried that first, but we need to support 18.04 too which doesn't have wget installed by default. It needs to call `apt-get update` which will fail with invalid key. My fix is the alternative approach from nvidia which hopefully works for both 18.04 and 20.04.

**Comment by Tixxx** [comment]
> @Tixxx this is how @tgaddair fixed it for Ludwig: [ludwig-ai/ludwig@1a4f679](https://github.com/ludwig-ai/ludwig/commit/1a4f67914bc3894c7ada67a5b985c9c1ca2a3cc0)
yea, I tried that first, but we need to support 18.04 too which doesn't have wget installed by default. It needs to call apt-get update which will fail with invalid key. My fix is the alternative approach from nvidia which hopefully works for both 18.04 and 20.04.

**Review by tgaddair — APPROVED** [review]

**Pull request merged**
